s3-source-connector/src/main/java/io/aiven/kafka/connect/s3/source/utils/SourceRecordIterator.java

```diff
@@ -0,0 +1,217 @@
+                defaultPartitionId = Integer.parseInt(fileMatcher.group(PATTERN_PARTITION_KEY));
+            } else {
+                LOGGER.error("File naming doesn't match to any topic. {}", currentObjectKey);
+                inputStream.abort();
+                s3Object.close();
+                return Collections.emptyIterator();
+            }
+
+            final long defaultStartOffsetId = 1L;
+
+            final String finalTopic = topicName;
+            final Map<String, Object> partitionMap = ConnectUtils.getPartitionMap(topicName, defaultPartitionId,
+                    bucketName);
+
+            return getObjectIterator(inputStream, finalTopic, defaultPartitionId, defaultStartOffsetId, transformer,
+                    partitionMap);
+        }
+    }
+
+    @SuppressWarnings("PMD.CognitiveComplexity")
+    private Iterator<S3SourceRecord> getObjectIterator(final InputStream valueInputStream, final String topic,
+            final int topicPartition, final long startOffset, final Transformer transformer,
+            final Map<String, Object> partitionMap) {
+        return new Iterator<>() {
+            private final Iterator<S3SourceRecord> internalIterator = readNext().iterator();
+
+            private List<S3SourceRecord> readNext() {
+                final byte[] keyBytes = currentObjectKey.getBytes(StandardCharsets.UTF_8);
+                final List<S3SourceRecord> sourceRecords = new ArrayList<>();
+
+                int numOfProcessedRecs = 1;
+                boolean checkOffsetMap = true;
+                for (final Object record : transformer.getRecords(valueInputStream, topic, topicPartition,
+                        s3SourceConfig)) {
+                    if (offsetManager.shouldSkipRecord(partitionMap, currentObjectKey, numOfProcessedRecs)
+                            && checkOffsetMap) {
+                        numOfProcessedRecs++;
+                        continue;
+                    }
+
+                    final byte[] valueBytes = transformer.getValueBytes(record, topic, s3SourceConfig);
+                    checkOffsetMap = false;
+                    sourceRecords.add(getSourceRecord(keyBytes, valueBytes, topic, topicPartition, offsetManager,
+                            startOffset, partitionMap));
+                    if (sourceRecords.size() >= s3SourceConfig.getInt(MAX_POLL_RECORDS)) {
+                        break;
+                    }
+
+                    numOfProcessedRecs++;
+                }
+                return sourceRecords;
+            }
+
+            private S3SourceRecord getSourceRecord(final byte[] key, final byte[] value, final String topic,
+                    final int topicPartition, final OffsetManager offsetManager, final long startOffset,
+                    final Map<String, Object> partitionMap) {
+
+                long currentOffset;
+
+                if (offsetManager.getOffsets().containsKey(partitionMap)) {
```

Comment:
Would it make sense to move this if-else into the offset manager, we can create an interface, and re-use this functionality in all source connectors going forward?